Add function `arrayRandomSample()`

itayisraelov opened the pull request:
Fixes #53098

```sql
SELECT arrayRandomSample([1,2,3], 2); -- [1,2] or [2,3] or [1,3]

SELECT arrayRandomSample([[1,2,3], [3,4], [6,9]], 2); -- [[1,2,3], [3,4]] or [[3,4], [6,9]] or ...
```

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added a new SQL function, "arrayRandomSample(arr, k)" which returns a sample of k elements from the input array. Similar functionality could previously be achieved only with less convenient syntax, e.g. "SELECT arrayReduce('groupArraySample(3)', range(10))".

itayisraelov committed:
implement randomSampleFromArray function

itayisraelov commented:
Hey @rschu1ze :slightly_smiling_face: ,
With which version of Clang are you working? clang-16 or clang-17?
because I have little problems with the linking step when using clang-17, and I want to validate that it is not my computer. Thanks :)

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #54391 from itayisraelov/ClickHouse:israelov/generate-random-int-array -->
This is an automated comment for commit 858cf51745dc0733673d109e3724eb82483c3e67 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54391/858cf51745dc0733673d109e3724eb82483c3e67/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🔴 failure</td></tr>
</table>

rschu1ze commented:
Hi @itayisraelov, CI uses clang-16, the [upgrade to clang-17](https://github.com/ClickHouse/ClickHouse/pull/53831) is wip.

itayisraelov committed:
Add SQL query tests by script shell

itayisraelov committed:
add documentation

rschu1ze commented on docs/en/sql-reference/functions/array-functions.md:
````diff
@@ -2118,6 +2118,113 @@ Result:
 ```
 
+
+## randomSampleFromArray
````
Basically *all* functions that process arrays start with `array`. We should rather name this function `arrayRandomSample()`.

itayisraelov replied:
It makes sense. I will fix it.

itayisraelov replied:
done

rschu1ze commented on docs/en/sql-reference/functions/array-functions.md:
````diff
@@ -2118,6 +2118,113 @@ Result:
+└─────────┘
+```
+
+The `randomSampleFromArray` function randomly selects elements from the input array, ensuring that the specified number of elements is included in the sample. If `sampleSize` exceeds the size of the input array, it will be limited to the size of the array. It can handle both flat arrays and arrays containing nested arrays, providing flexibility in sampling from complex data structures.
````
This sentence should rather appear somewhere around l. 2124. But perhaps we don't need it at all since all it says has been said above already.

itayisraelov replied:
removed

rschu1ze commented on docs/en/sql-reference/functions/array-functions.md:
```diff
@@ -2118,6 +2118,113 @@ Result:
+**Arguments**
+
+- `arr` — The input array from which to sample elements, including nested arrays.
+- `sampleSize` — The number of elements to include in the random sample.
```
Does it make sense to include an optional seed parameter for reproducibility?

itayisraelov replied:
This is a good idea. I will check it :)

itayisraelov replied:
As we agreed, this is nice to have. Maybe it will be implemented in the future.

rschu1ze commented on docs/en/sql-reference/functions/array-functions.md:
```diff
@@ -2118,6 +2118,113 @@ Result:
+┌─res─────┐
+│ [2,3,1] │
+└─────────┘
+or
```
Including a single example result would be just fine.

itayisraelov replied:
removed

rschu1ze commented on docs/en/sql-reference/functions/array-functions.md:
````diff
@@ -2118,6 +2118,113 @@ Result:
+
+Query:
+
+```sql
````
The example is conceptually the same as the previous one. We can omit one of these. In case of doubt, please go for the second one (it is a little bit more illustrative).

itayisraelov replied:
done

rschu1ze commented on docs/en/sql-reference/functions/array-functions.md:
```diff
@@ -2118,6 +2118,113 @@ Result:
+**Arguments**
+
+- `arr` — The input array from which to sample elements, including nested arrays.
+- `sampleSize` — The number of elements to include in the random sample.
```
Haven't looked at the code yet, but you want to specify the exact data types here. In particular, `sampleSize` could be made an unsigned integer. This would prevent the the case that it is negative (in which case there are multiple, not immediately intuitive options, e.g. 1. return the input array as is, 2. return an empty array, return NULL, 4. unspecified etc.)

itayisraelov replied:
done

rschu1ze replied:
Could we document also here, that sampleSize is an unsigned integer?

itayisraelov replied:
I wrote "done", but then pushes the fix with some delay. This is probably the reason that you wasn't saw it. I hope now it is OK.

rschu1ze commented on docs/en/sql-reference/functions/array-functions.md:
````diff
@@ -2118,6 +2118,113 @@ Result:
+
+Result:
+```
+TODO: FIX ME
````
Return an empty array?

itayisraelov replied:
Yes, but when I am running this query:

`clickhouse-client --query "SELECT randomSampleFromArray([[1,2,3], [3,4], [6,9]], 0);`

I am getting this error:

```
Received exception from server (version 23.9.1):
Code: 9. DB::Exception: Received from localhost:9000. DB::Exception: Incorrect size of nested column in constructor of ColumnConst: 0, must be 1.: While processing randomSampleFromArray([[1, 2, 3], [3, 4], [6, 9]], 0). (SIZES_OF_COLUMNS_DOESNT_MATCH)
(query: SELECT randomSampleFromArray([[1,2,3], [3,4], [6,9]], 0);
)
```

I will investigate it.

itayisraelov replied:
I tried it, but the exception did not come from the execute method.

rschu1ze replied:
Function arguments can generally be const or non-const. For example, a const string and a non-const string have different internal representations. With one function parameter, that's 2 variants to implement, with two function parameters, it's 4, with three function parameters, it's 8 etc. Usually, not all combinations need to be implemented as certain function parameters are usually only const (in this case, the sampleCount parameter is always const).

If `useDefaultImplementationForConstants = true`, then ClickHouse converts an invocation with only const parameters into an invocation with only non-const parameters. This basically happens with `arrayRandomSample(['a', 'b', 'c'], 2)`. The motivation for that is to reduce the number of variants that the programmer needs to implement. The problem is that something goes wrong during the conversion here (didn't check in detail).

We can look at the way function `arrayRotateLeft` is implemented. Turns out that they set  `useDefaultImplementationForConstants = false` (i.e. inherit the default by not overriding the virtual method) and implement the const/non-const separately. Maybe that's an option for you too.

itayisraelov replied:
Thanks for the detailed answer :slightly_smiling_face: . It was helpful information for me. 
I will take a look.

itayisraelov replied:
In the end, this line solved the problem:
```
if (K == 0)
  {
      // Handle edge cases where input array is empty or K is 0
      res_data->getOffsets().push_back(0);
      return res_data;
  }
```

rschu1ze replied:
Sounds good. Let me quickly verify locally, we can merge then.

(the optional seed parameter is perhaps overkill - it will only be useful for testing but your tests cater for randomness quite well.)

itayisraelov replied:
OK :) 
I still need to iterate on all your open comments and solve them.
And one more case that I need to check:
Currently, my code does not support multi- rows implementation. 
I will solve it (including tests), and then I will ping you.

rschu1ze replied:
Unrelated comment: Could you please chmod +x to the `.sh` test file. Otherwise, `clickhouse-test` ([docs](https://clickhouse.com/docs/en/development/tests#functional-test-locally)) complains for me about permissions.

rschu1ze replied:
It also looks like you would need to re-generate the expected result file. It is out-of-sync with the test.

rschu1ze replied:
What doesn't work yet are
- `select arrayRandomSample(['a', 'b', 'c'], materialize(1))` and
- `select arrayRandomSample(materialize(['a', 'b', 'c']), 1)`

itayisraelov replied:
Yes, exactly, I saw it. 
2 minutes. 
Thanks!

itayisraelov replied:
I have not succeeded in treating all comments yet. One of them is the "materialize" comment. 
Now, I will take a look.

itayisraelov replied:
The empty array problem - solved
The materialize function is blocking until we fix it with flow-up PR.

rschu1ze replied:
Unfortuantely, the materialize problem blocks merge. It is relevant for 99% of all practical use cases, basically whenever the function is applied to a column of arrays. Not supporting it properly raise eyebrows with many users.

But no hurry to finish this PR up. Either you can do it in 2, 3, ... weeks, or if not let me know, I can also finish it (when I find some time).

itayisraelov replied:
Yes, you are right; I will fix it with this PR.

rschu1ze commented on docs/en/sql-reference/functions/array-functions.md:
````diff
@@ -2118,6 +2118,113 @@ Result:
+
+Query:
+
+```sql
````
Let's omit the example for this special case, it's clear what happens anyways.

rschu1ze replied:
Ah, I missed that the results are in different order, potentially. Then keep the example - thanks.

rschu1ze commented on docs/en/sql-reference/functions/array-functions.md:
```diff
@@ -2118,6 +2118,113 @@ Result:
+**Arguments**
+
+- `arr` — The input array from which to sample elements, including nested arrays.
+- `sampleSize` — The number of elements to include in the random sample.
```
Let's rename `sampleSize` to `samples`.

itayisraelov replied:
done

rschu1ze commented on docs/en/sql-reference/functions/array-functions.md:
```diff
@@ -2118,6 +2118,113 @@ Result:
+
+## randomSampleFromArray
+
+Returns a random sample from an input array. The number of elements in the sample is determined by the `sampleSize` argument. The function ensures that each element in the input array has an equal chance of being included in the sample.
```
(I'd probably omit the last sentence, it is not wrong but it states the obvious.)

rschu1ze commented on docs/en/sql-reference/functions/array-functions.md:
```diff
@@ -2118,6 +2118,113 @@ Result:
+
+**Arguments**
+
+- `arr` — The input array from which to sample elements, including nested arrays.
```
Let's omit the part after the comma, it is confusing for readers.

itayisraelov replied:
removed

rschu1ze commented on tests/queries/1_stateful/00178_random_sample_from_array.sh:
```diff
@@ -0,0 +1,62 @@
+
+# Test randomSampleFromArray with integer array
+query_result=$(clickhouse-client -q "SELECT randomSampleFromArray([1,2,3], 2)")
+if [[ "$query_result" =~ "1" ]] || [[ "$query_result" =~ "2" ]] || [[ "$query_result" =~ "3" ]]; then
```
So, if the function happens to return `[1]`, this will count as passed, right? In this case, better test against all three possible result arrays `[1,2]`, `[1,3]` and `[2,3]`.

itayisraelov replied:
The tests had a bug; the checks were incorrect, so I fixed it. You now should see different checks.

rschu1ze commented on src/Functions/array/randomSampleFromArray.cpp:
```diff
@@ -0,0 +1,126 @@
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        const DataTypeArray * array_type = checkAndGetDataType<DataTypeArray>(arguments[0].get());
```
You might want to look into `FunctionArgumentDescriptors` and `validateFunctionArgumentTypes()` to validate the input types. They are more high-level than the current code.

itayisraelov replied:
Thanks, I will take a look.

itayisraelov replied:
I tried today once, but the function is not recognized. 
I will try it again now.

itayisraelov replied:
done

rschu1ze commented on src/Functions/array/randomSampleFromArray.cpp:
```diff
@@ -0,0 +1,126 @@
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t) const override
+    {
+        const ColumnArray * input_data = checkAndGetColumn<ColumnArray>(arguments[0].column.get());
```
`input_data` --> `column_array` or `col_array`

itayisraelov replied:
done

rschu1ze commented on tests/queries/1_stateful/00178_random_sample_from_array.sh:
```diff
@@ -0,0 +1,62 @@
+    echo "sorted_result: ${sorted_result[*]}"
+fi
+((total_tests++))
+
```
You might want to include tests for the cases that the array or the sample count are non-const. See SQL function `materialize()` for that.

(To simply things, I think it makes sense to require that the sample count argument is const. You can check that in getReturnTypeImpl().)

itayisraelov replied:
No problem, I will add a test for non-const samples.
I already added this check:
```
if (!isUnsignedInteger(arguments[1]))
            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Second argument must be a unsigned integer");
```

rschu1ze replied:
Yes. But not only check but throw an exception if it isn't. Non-const `samples` are an edge case which complicate the implementation.

itayisraelov replied:
How to solve the problem for the array param?

rschu1ze replied:
Take a look at the implementation of `arrayRotateLeft`. The code needs to be able to handle const and non-const array parameters.

rschu1ze commented:
@itayisraelov Oh, and thanks for your contribution. This is a promising PR!

itayisraelov committed:
fix comments: rname randomSampleFromArray --> arrayRandomSample

rschu1ze commented on utils/check-style/aspell-ignore/en/.aspell.en.prepl:
```diff
@@ -0,0 +1 @@
+personal_repl-1.1 en 0 
```
File committed unintentionally?

itayisraelov replied:
Hmm, let me check.

itayisraelov replied:
Removed, thanks!

itayisraelov committed:
fix comments 2

itayisraelov commented on src/Functions/array/randomSampleFromArray.cpp:
```diff
@@ -0,0 +1,126 @@
+#include <random>
```
Yep, I wanted to do it, and I forgot. Thanks.

itayisraelov commented:
> @itayisraelov Oh, and thanks for your contribution. This is a promising PR!

Thanks for the help here :slightly_smiling_face:

rschu1ze commented on docs/en/sql-reference/functions/array-functions.md:
````diff
@@ -2118,6 +2118,80 @@ Result:
+
+Result:
+```
+TODO: FIX ME
````
(still says FIX ME)

itayisraelov replied:
Yep, please give me a couple of minutes. I will fix everything.

rschu1ze commented on src/Functions/array/randomSampleFromArray.cpp:
```diff
@@ -0,0 +1,126 @@
+        std::random_device rd;
+        std::mt19937 gen(rd());
+
+        if (static_cast<UInt64>(num_elements) < K)
```
Should be sth like: K = std::min(K, num_elements)

itayisraelov replied:
Yep, it is also acceptable :slightly_smiling_face:

itayisraelov replied:
done

itayisraelov commented on utils/check-style/aspell-ignore/en/aspell-dict.txt:
```diff
@@ -1,4 +1,4 @@
-personal_ws-1.1 en 543
+personal_ws-1.1 en 2586 
```
Do you think this change is OK?

rschu1ze replied:
So this seems to change the encoding. What did 543 stand for, what does 2586 stand for? To be on the safe side, please leave this change out or push it in a separate PR.

itayisraelov replied:
Removed, it works without it. Thanks!

itayisraelov committed:
fix comments 3

itayisraelov committed:
clang-format

itayisraelov committed:
fix empty array bug

itayisraelov committed:
add more tests
1. when k=0
2. when array is empty && k>0
3. when k is negative, string or floating point.

itayisraelov commented on src/Functions/array/arrayRandomSample.cpp:
```diff
@@ -0,0 +1,143 @@
+
+            column_array_ptr = column_const_array->convertToFullColumn();
+            column_array = assert_cast<const ColumnArray *>(column_array_ptr.get());
+        }
```
@rschu1ze, these codes fix the materialize function option, but now I see two rows in the result for some reason.
I am checking it. 

See this one for example:
```
home :) select arrayRandomSample(['a', 'b', 'c'], materialize(1));

SELECT arrayRandomSample(['a', 'b', 'c'], materialize(1))

Query id: 4b497658-8b39-4bcf-b4af-64a2f6137c6e

┌─arrayRandomSample(['a', 'b', 'c'], materialize(1))─┐
│ []                                                 │
└────────────────────────────────────────────────────┘
┌─arrayRandomSample(['a', 'b', 'c'], materialize(1))─┐
│ ['b']                                              │
└────────────────────────────────────────────────────┘

2 rows in set. Elapsed: 0.004 sec. 
```

rschu1ze replied:
Weird. `select arrayRandomSample(['a', 'b', 'c'], materialize(1));` should throw an exception.

itayisraelov replied:
I suspect these lines:
```
column_array_ptr = column_const_array->convertToFullColumn();
column_array = assert_cast<const ColumnArray *>(column_array_ptr.get());
```

itayisraelov replied:
In the end, I do not block samples to by only const, block it? 
I did the same code as array param:
```
ColumnPtr col_num_ptr = arguments[1].column;
        const IColumn * col_num = col_num_ptr.get();

        if (!col_num)
        {
            const auto * column_const_num = checkAndGetColumnConst<ColumnConst>(col_num_ptr.get());
            if (!column_const_num || column_const_num->empty())
                throw Exception(
                    ErrorCodes::ILLEGAL_COLUMN, "The second argument column is empty or null, type = {}", arguments[1].type->getName());
            col_num_ptr = column_const_num->convertToFullColumn();
            col_num = col_num_ptr.get();
        }
```

itayisraelov replied:
I am getting some errors; let's solve this problem with a separate PR. 
because my time is a little limited for the following week.

itayisraelov committed:
Support multi-row implementation, but samples is const

itayisraelov committed:
add tests for multi-row

itayisraelov committed:
fix comments 4 - fix getReturnTypeImpl implemention impl
use FunctionArgumentDescriptors && validateFunctionArgumentTypes

rschu1ze committed:
Fixups

rschu1ze committed:
Merge remote-tracking branch 'rschu1ze/master' into israelov/generate-random-int-array

rschu1ze committed:
Suppress some shellcheck errors

rschu1ze committed:
Try to fix flakyness

rschu1ze committed:
Try to fix flakyness, pt. II

rschu1ze commented:
Stress test failures: #55151

alexey-milovidov commented on src/Functions/array/arrayRandomSample.cpp:
```diff
@@ -0,0 +1,118 @@
+        }
+
+        std::random_device rd;
+        std::mt19937 gen(rd());
```
@rschu1ze this is wrong.

Mersenne-Twister is an outdated PRNG that should not be used.

alexey-milovidov commented on src/Functions/array/arrayRandomSample.cpp:
```diff
@@ -0,0 +1,118 @@
+        {
+            samples = col_samples->getUInt(0);
+        }
+        catch (...)
```
@rschu1ze This is wrong. We should not catch everything.
We are losing the original stack here.

alexey-milovidov commented on src/Functions/array/arrayRandomSample.cpp:
```diff
@@ -0,0 +1,118 @@
+        {
+            size_t row_size = input_offsets[row] - current_offset;
+
+            std::vector<size_t> indices(row_size);
```
This is wrong. It is ridiculously inefficient.

alexey-milovidov commented:
@rschu1ze, please remove all usages of Mersenne Twister from the codebase.
Then add a style check preventing its usage.
